language/vm/vm-runtime/src/process_txn/execute.rs

```diff
@@ -66,57 +60,44 @@ where
             )
         }
         TransactionPayload::Module(module) => {
-            let VerifiedTransactionState {
-                mut txn_executor,
-                verified_txn,
-            } = txn_state.expect("module transactions should always have associated state");
-            let ver_module = match verified_txn {
-                VerTxn::Module(ver_module) => ver_module,
-                _ => unreachable!("TransactionPayload::Module expects VerTxn::Module"),
-            };
-            let module_id = ver_module.self_id();
-            // Make sure that there is not already a module with this name published
-            // under the transaction sender's account.
-            // Note: `exists_module` will fetch the module from either the
-            //       global storage or from the local data cache (which means that this module is
-            //       published within the same block).
-            if txn_executor.exists_module(&module_id) {
-                return txn_executor.failed_transaction_cleanup(Err(vm_error(
-                    Location::default(),
-                    StatusCode::DUPLICATE_MODULE_NAME,
-                )));
+            let module = module.into_inner();
+            let module_id = match executor.publish_module(&module, runtime, state_view) {
+                Ok(module_id) => module_id,
+                Err(err) => match err.status_type() {
+                    StatusType::InvariantViolation => {
+                        error!("[VM] VM error running script: {:?}", err);
+                        return ExecutedTransaction::discard_error_output(err);
+                    }
+                    _ => {
+                        warn!("[VM] User error running script: {:?}", err);
+                        return executor.failed_transaction_cleanup(runtime, state_view, Err(err));
+                    }
+                },
             };
-            let module_bytes = module.into_inner();
-            let output = txn_executor.transaction_cleanup(vec![(module_id, module_bytes)]);
+
+            executor.transaction_cleanup(runtime, state_view, vec![(module_id, module)])
+            // REVIEW: should this go in ever? essentially should a module be published eagerly?
```

Comment:
I'd like to leave this here as is and think some more about it.
I am not worried about this sticking in the code because this logic is going to be sorted out next

Reply:
IMHO with our current setup, we should uncomment this block of code right now as the code cache is only hooked up with StateView rather than the BlockDataCache. Thus we have to publish it eagerly so that later transaction can see this new module.

Reply:
isn't the case that later transaction will ask the data view for the new code and fetch it that way?
I mean this is publishing the code which may be fine but I wonder if we want it this way.
It seems to me at the end of a publish transaction we publish the code explicitly in the write set. Isn't that the case?

Reply:
ok if I understand I'll put up a TODO here and remove the code